agent/secret.go

```diff
@@ -0,0 +1,50 @@
+package agent
+
+import (
+	"strings"
+
+	"github.com/drone/drone/model"
+)
+
+// SecretReplacer hides secrets from being exposed by the build output.
+type SecretReplacer interface {
+	// Replace conceals instances of secrets found in s.
+	Replace(s string) string
+}
+
+// NewSecretReplacer creates a SecretReplacer based on whether any value in
+// secrets requests it be hidden.
+func NewSecretReplacer(secrets []*model.Secret) SecretReplacer {
+	var r []string
+	for _, s := range secrets {
+		if s.Conceal {
+			r = append(r, s.Value, "*****")
+		}
+	}
+
+	var replacer SecretReplacer
+
+	if len(r) > 0 {
+		replacer = &secretReplacer{
+			replacer: strings.NewReplacer(r...),
+		}
+	} else {
+		replacer = &noopReplacer{}
+	}
+
+	return replacer
+}
```

Comment:
more of a styling note, but go style guidelines would say to drop the if / else and instead do the following:

```
if len(r) == 0 {
  return &noopReplacer{}
}
return replacer = &secretReplacer{
  replacer: strings.NewReplacer(r...),
}
```